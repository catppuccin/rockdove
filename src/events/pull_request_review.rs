use octocrab::models::{
    pulls::ReviewState,
    webhook_events::{
        payload::{PullRequestReviewWebhookEventAction, PullRequestReviewWebhookEventPayload},
        WebhookEvent,
    },
};

use crate::{
    colors::PULL_REQUEST_COLOR,
    embed_builder::EmbedBuilder,
    errors::{RockdoveError, RockdoveResult},
};

pub fn make_embed(
    event: WebhookEvent,
    specifics: &PullRequestReviewWebhookEventPayload,
) -> RockdoveResult<Option<EmbedBuilder>> {
    if !matches!(
        specifics.action,
        PullRequestReviewWebhookEventAction::Submitted
    ) {
        return Ok(None);
    }

    let repo = event
        .repository
        .ok_or_else(|| RockdoveError::MissingField {
            event_type: event.kind.clone(),
            field: "repository",
        })?;

    let mut embed = EmbedBuilder::default();

    let repo_name = repo.full_name.unwrap_or(repo.name);

    embed.title(&format!(
        "[{}] Pull request {}: #{} {}",
        repo_name,
        match specifics
            .review
            .state
            .ok_or_else(|| RockdoveError::MissingField {
                event_type: event.kind.clone(),
                field: "review.state",
            })? {
            ReviewState::Approved => "approved",
            ReviewState::ChangesRequested => "changes requested",
            ReviewState::Commented => "reviewed",
            _ => return Ok(None),
        },
        specifics.pull_request.number,
        specifics
            .pull_request
            .title
            .as_ref()
            .ok_or_else(|| RockdoveError::MissingField {
                event_type: event.kind.clone(),
                field: "pull_request.title",
            })?
    ));

    embed.url(specifics.review.html_url.as_str());

    if let Some(ref body) = specifics.review.body {
        embed.description(body);
    }

    embed.color(PULL_REQUEST_COLOR);

    Ok(Some(embed))
}

#[cfg(test)]
mod tests {
    use crate::snapshot_test;

    use yare::parameterized;

    #[parameterized(
        approved = { "approved" },
        changes_requested = { "changes_requested" },
        commented = { "commented" }
      )]
    fn snapshot(event_type: &str) {
        snapshot_test!("pull_request_review", event_type);
    }
}
