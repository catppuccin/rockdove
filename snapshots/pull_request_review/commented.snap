---
source: src/events/pull_request_review.rs
info:
  author_name_length: 8
  colour_hex: "#89B4FA"
  description_length: 15
  title_length: 76
---
embeds:
  - author:
      icon_url: "https://avatars.githubusercontent.com/u/58985301?v=4"
      name: sgoudham
      url: "https://github.com/sgoudham"
    color: 9024762
    description: a normal review
    title: "[catppuccin-rfc/polybar] Pull request reviewed: #3 chore: Configure Renovate"
    url: "https://github.com/catppuccin-rfc/polybar/pull/3#pullrequestreview-2319885438"
